sw/host/opentitanlib/src/rom_patch/patch.rs

```diff
@@ -0,0 +1,745 @@
+    }
+}
+
+impl<'a> Read for RomPatch<'a> {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        if buf.len() < self.size {
+            return Err(std::io::Error::from(std::io::ErrorKind::InvalidInput));
+        }
+
+        let mut buf_offset = self.header.read(buf)?;
+        let match_table_size = PATCH_MATCH_TABLE_SIZE;
+        let raw_match_table = &self.match_table as *const _ as *const u8;
+        unsafe {
+            ptr::copy_nonoverlapping(
+                raw_match_table,
+                buf[buf_offset..].as_mut_ptr(),
+                match_table_size,
+            );
+        }
+
+        /* Fill the rest of the patch with NOPs */
+        let nop_bytes = RV32_NOP.to_le_bytes();
+        let mut offset = 0;
+        let code_section = &mut buf[self.code_offset()..self.code_offset() + self.code_len()];
+        while offset + 4 < code_section.len() {
+            unsafe {
+                ptr::copy_nonoverlapping(
+                    nop_bytes.as_ptr(),
+                    code_section[offset..].as_mut_ptr(),
+                    4,
+                );
+            }
+
+            offset += 4;
+        }
+
+        /* Copy the whole patch section */
+        buf_offset = self.code_offset();
+        for insn in &self.patch_code {
+            let insn_bytes = insn.to_le_bytes();
+            let raw_insn_bytes = &insn_bytes as *const _ as *const u8;
+            unsafe {
+                ptr::copy_nonoverlapping(raw_insn_bytes, buf[buf_offset..].as_mut_ptr(), 4);
+            }
+            buf_offset += 4;
+        }
+
+        /* Copy all subroutines, at their respective offsets */
+        for sub_routine in self.sub_routines.iter_mut() {
+            if sub_routine.read(&mut buf[sub_routine.offset()..])? != sub_routine.len() {
+                return Err(std::io::Error::new(
+                    io::ErrorKind::Other,
+                    "Subroutine read failure".to_string(),
+                ));
+            }
+        }
+
+        /*
+         * Compute the patch signature.
+         * The first byte (The header meta data, lock valid and program start)
+         * is not part of the signed payload as it can be toggled by the patch
+         * loader.
+         */
+        let signed_payload = &buf[offset_of!(RomPatchHeader, size)..self.signature_offset()];
+        let digest = sha256::sha256(signed_payload);
+        let signature: &[u8] = &self
+            .signature_key
+            .sign(&digest)
+            .map_err(|e| std::io::Error::new(io::ErrorKind::Other, e.to_string()))?
+            .to_le_bytes()[..];
+
+        /* Copy the signature */
+        unsafe {
+            ptr::copy_nonoverlapping(
+                signature.as_ptr(),
+                buf[self.signature_offset()..].as_mut_ptr(),
+                RSA_SIGNATURE_LEN_BYTES,
+            );
+        }
+
+        Ok(self.len())
+    }
+}
+
+pub struct RomPatchPartition<'a> {
+    size: usize,
```

Comment:
size of what?

Reply:
Not relevant anymore with the updated PR.